Templates: Add back button & fix focus loss when navigating through template creation flow

Mayank-Tripathi32 opened the pull request:
Added by @t-hamano 

> [!NOTE]
> Don't forget to add @Sourav61, who worked in #69711,  as a co-author when merging this PR.

---

## What?
Closes #69711

Adds a Back button to the template creation flow and sets focus on the name input field when creating a custom template.

## Why?
When clicking the "Custom template" button, the focus is currently lost, causing accessibility issues for screen reader users. This PR fixes the focus management and improves the navigation flow by adding a Back button that allows users to return to the previous step without closing the entire modal.

## How?
- Added a Back button to the custom template modal that returns users to the template selection screen
- Implemented focus management using requestAnimationFrame to set focus on the name input field when the custom template modal opens
- Removed the redundant Cancel button to simplify the UI and make navigation more intuitive

## Testing Instructions
1. Open the Site Editor
2. Click on "Templates" in the navigation menu
3. Click the "Add Template" button
4. Click the "Custom template" button
5. Verify that focus is set on the Name input field
6. Click the "Back" button
7. Verify that you return to the template selection screen

### Testing Instructions for Keyboard
1. Open the Site Editor
2. Use Tab to navigate to the "Templates" option in the navigation menu and press Enter
3. Tab to the "Add Template" button and press Enter
4. Tab to the "Custom template" button and press Enter
5. Verify that focus is automatically set on the Name input field
6. Tab to the "Back" button and press Enter
7. Verify that you return to the template selection screen

## Screencast
https://q7utzrengv.ufs.sh/f/Wgl9eBAmTj29r0ltzJ9bvXLPnVf0wHAxqleM96uGDdWoCa3T

Mayank-Tripathi32 committed:
feat: add focus to name input and back navigation

Mayank-Tripathi32 commented:
I would like some feedback on PR and on issue with other focus loss:

Here is video, that better shows the focus loss when going to post based templates
https://q7utzrengv.ufs.sh/f/Wgl9eBAmTj29DqvG0WCE9O3LjigMnKlJx2AvwtoUc0hWd7yz

Let me know if we should fix these in this PR itself

t-hamano commented on packages/edit-site/src/components/add-new-template/add-custom-generic-template-modal-content.js:
```diff
@@ -15,10 +15,24 @@ import {
+	// Set focus to the name input when the component mounts
+	useEffect( () => {
+		// Use requestAnimationFrame to ensure the DOM is fully rendered
+		const frameId = window.requestAnimationFrame( () => {
+			if ( inputRef.current ) {
+				inputRef.current.focus();
+			}
+		} );
+
+		return () => window.cancelAnimationFrame( frameId );
```
```suggestion
		if ( inputRef.current ) {
			inputRef.current.focus();
		}
```

Do we need to use rAF?

Mayank-Tripathi32 replied:
Yes, it's better to use requestAnimationFrame instead of calling .focus() directly, as the input element might not be fully mounted or rendered yet.

Through, rAF itself is not very reliable. So I can't say for certain.

t-hamano replied:
Isn't useEffect executed _after_ the component is mounted? If the element with ref were to be conditionally rendered based on some state, rAF might be necessary, but for this component, I think it's guaranteed that the element with ref applied will always be rendered.

Mayank-Tripathi32 replied:
You're right, it's essentially a safeguard, but I think it's safe to remove it in this case.

t-hamano reviewed (commented):
Thanks for the PR! I think this PR is a great a11y improvement, but we need to address two more things:


First, we need the back button in the "Add template: Post" modal as well:

![image](https://github.com/user-attachments/assets/13b2ccf1-ee94-4f4c-bc76-f407feb1ef14)

Second, when the Back button is pressed, focus is lost. We should focus either the modal itself or the first focusable element within the modal.

Mayank-Tripathi32 committed:
refactor: simplify focus handling for name input in template modal

Mayank-Tripathi32 committed:
feat: add back navigation functionality to custom template modal

Mayank-Tripathi32 committed:
feat: enhance focus management in create template modals

Mayank-Tripathi32 commented:
Hey @t-hamano, I have added the back button in all steps and fixed focus loss issue.

https://q7utzrengv.ufs.sh/f/Wgl9eBAmTj29pojunySmx9PaNdTZpH58C7GnvcBY3RjewJus

Can you check? Also should we focus heading or first tab item? I feel that currently it just loops 3 times "Add template" when creating posts/page templates.

t-hamano commented on packages/edit-site/src/components/add-new-template/add-custom-template-modal-content.js:
```diff
@@ -165,10 +166,27 @@ function SuggestionList( { entityForSuggestions, onSelect } ) {
 		entityForSuggestions.hasGeneralTemplate
 	);
+
+	// Focus on the first focusable element when the modal opens.
+	// We handle focus management in the parent modal, just need to focus on the first focusable element.
+	useEffect( () => {
+		if ( containerRef.current ) {
+			const [ firstFocusable ] = focus.focusable.find(
+				containerRef.current
+			);
+			firstFocusable?.focus();
+		}
+	}, [ showSearchEntities ] );
```
I'm thinking that we could centralize all focus management in the modal, i.e. `NewTemplateModal`. I tried the following code.

<details><summary>Details</summary>

```diff
diff --git a/packages/edit-site/src/components/add-new-template/add-custom-generic-template-modal-content.js b/packages/edit-site/src/components/add-new-template/add-custom-generic-template-modal-content.js
index dd8ccce86f..e5b21c8633 100644
--- a/packages/edit-site/src/components/add-new-template/add-custom-generic-template-modal-content.js
+++ b/packages/edit-site/src/components/add-new-template/add-custom-generic-template-modal-content.js
@@ -6,7 +6,7 @@ import { paramCase as kebabCase } from 'change-case';
 /**
  * WordPress dependencies
  */
-import { useState, useEffect, useRef } from '@wordpress/element';
+import { useState, useRef } from '@wordpress/element';
 import { __ } from '@wordpress/i18n';
 import {
        Button,
@@ -21,13 +21,6 @@ function AddCustomGenericTemplateModalContent( { createTemplate, onBack } ) {
        const [ isBusy, setIsBusy ] = useState( false );
        const inputRef = useRef();
 
-       // Set focus to the name input when the component mounts
-       useEffect( () => {
-               if ( inputRef.current ) {
-                       inputRef.current.focus();
-               }
-       }, [] );
-
        async function onCreateTemplate( event ) {
                event.preventDefault();
                if ( isBusy ) {
diff --git a/packages/edit-site/src/components/add-new-template/add-custom-template-modal-content.js b/packages/edit-site/src/components/add-new-template/add-custom-template-modal-content.js
index 7a846985c8..3b0d9b998c 100644
--- a/packages/edit-site/src/components/add-new-template/add-custom-template-modal-content.js
+++ b/packages/edit-site/src/components/add-new-template/add-custom-template-modal-content.js
@@ -16,7 +16,6 @@ import {
 import { useEntityRecords } from '@wordpress/core-data';
 import { decodeEntities } from '@wordpress/html-entities';
 import { useDebouncedInput } from '@wordpress/compose';
-import { focus } from '@wordpress/dom';
 
 /**
  * Internal dependencies
@@ -170,23 +169,11 @@ function AddCustomTemplateModalContent( {
        onSelect,
        entityForSuggestions,
        onBack,
-       containerRef,
 } ) {
        const [ showSearchEntities, setShowSearchEntities ] = useState(
                entityForSuggestions.hasGeneralTemplate
        );
 
-       // Focus on the first focusable element when the modal opens.
-       // We handle focus management in the parent modal, just need to focus on the first focusable element.
-       useEffect( () => {
-               if ( containerRef.current ) {
-                       const [ firstFocusable ] = focus.focusable.find(
-                               containerRef.current
-                       );
-                       firstFocusable?.focus();
-               }
-       }, [ showSearchEntities ] );
-
        return (
                <VStack
                        spacing={ 4 }
diff --git a/packages/edit-site/src/components/add-new-template/index.js b/packages/edit-site/src/components/ad
d-new-template/index.js
index aa36fa2e59..5ad4f11ff0 100644
--- a/packages/edit-site/src/components/add-new-template/index.js
+++ b/packages/edit-site/src/components/add-new-template/index.js
@@ -184,10 +184,7 @@ function NewTemplateModal( { onClose } ) {
        useEffect( () => {
                // Focus the first focusable element when component mounts or UI changes
                // We don't want to focus on the other modals because they have their own focus management.
-               if (
-                       containerRef.current &&
-                       modalContent === modalContentMap.templatesList
-               ) {
+               if ( containerRef.current ) {
                        const [ firstFocusable ] = focus.focusable.find(
                                containerRef.current
                        );
@@ -339,7 +336,6 @@ function NewTemplateModal( { onClose } ) {
                                        onBack={ () =>
                                                setModalContent( modalContentMap.templatesList )
                                        }
-                                       containerRef={ containerRef }
                                />
                        ) }
                        { modalContent === modalContentMap.customGenericTemplate && (
```

</details> 

This ensures that the modal dialog itself consistently gets focus when the modal content is switched. What do you think?

Mayank-Tripathi32 replied:
There are two reasons, I did it this way:

1. We wanted to focus on input field when its present, its related to this comment.
> In my opinion, the solution to this would be to set focus to the "Name" input, and also add a 'Back' option, so that a user can cancel the Create Custom Template action without having to start over. This should be true for all templates: selecting "Category Archives", for example, should provide an option to return to the selection screen.

2. This works fine but we will then need to move state of 
```JS
	const [ showSearchEntities, setShowSearchEntities ] = useState(
		entityForSuggestions.hasGeneralTemplate
	);
```
to parent component, as there is also focus loss when going from "Select whether to create a single template for all items or a specific one." to "This template will be used only for the specific item chosen." and back. This happens as useEffect does not have it as dependancy in parent block.

t-hamano replied:
I see, in the `AddCustomTemplateModalContent` component, the content switches depending on the `showSearchEntities` state, so this focus management is necessary.

t-hamano reviewed (commented):
Thanks for the PR! It's working very well so far.

t-hamano reviewed (approved):
In my testing, it works very well and is a great accessibility improvement.
